relay-server/src/actors/project_cache.rs

```diff
@@ -372,102 +375,70 @@ struct UpdateProjectState {
     no_cache: bool,
 }
 
-#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Copy)]
-struct QueueKey {
-    own_key: ProjectKey,
-    sampling_key: ProjectKey,
-}
-
-impl QueueKey {
-    fn new(own_key: ProjectKey, sampling_key: ProjectKey) -> Self {
-        Self {
-            own_key,
-            sampling_key,
-        }
-    }
-}
-
-/// The queue (buffer) of the incoming envelopes.
-#[derive(Debug, Default)]
-struct Queue {
-    /// Contains the cache of the incoming envelopes.
-    buffer: BTreeMap<QueueKey, Vec<(Box<Envelope>, EnvelopeContext)>>,
+/// Main broker of the [`ProjectCacheService`].
+///
+/// This handles incoming public messages, merges resolved project states, and maintains the actual
+/// cache of project states.
+#[derive(Debug)]
+struct ProjectCacheBroker {
+    config: Arc<Config>,
+    // Need hashbrown because drain_filter is not stable in std yet.
+    projects: hashbrown::HashMap<ProjectKey, Project>,
+    garbage_disposal: GarbageDisposal<Project>,
+    source: ProjectSource,
+    state_tx: mpsc::UnboundedSender<UpdateProjectState>,
     /// Index of the buffered project keys.
+    buffer_tx: mpsc::UnboundedSender<(Box<Envelope>, EnvelopeContext)>,
     index: BTreeMap<ProjectKey, BTreeSet<QueueKey>>,
+    buffer: Addr<Buffer>,
 }
 
-impl Queue {
-    /// Creates an empty queue.
-    pub fn new() -> Self {
-        Self::default()
-    }
-
+impl ProjectCacheBroker {
     /// Adds the value to the queue for the provided key.
     pub fn enqueue(&mut self, key: QueueKey, value: (Box<Envelope>, EnvelopeContext)) {
         self.index.entry(key.own_key).or_default().insert(key);
         self.index.entry(key.sampling_key).or_default().insert(key);
-        self.buffer.entry(key).or_default().push(value);
+        self.buffer.send(Enqueue::new(key, value));
     }
 
-    /// Returns the list of buffered envelopes if they satisfy a predicate.
-    pub fn dequeue<P>(
-        &mut self,
-        partial_key: &ProjectKey,
-        predicate: P,
-    ) -> Vec<(Box<Envelope>, EnvelopeContext)>
-    where
-        P: Fn(&QueueKey) -> bool,
-    {
+    /// Sends the message to [`BufferService`] to dequeue the envelopes.
```

Comment:
_nit_: Describe that the envelopes come back via the `buffer_rx` and are sent directly to `handle_processing`. This will make it easier for us to trace the control flow.